src/app/core/auth/auth.guard.ts

```diff
@@ -46,19 +51,78 @@ export class AuthGuard implements CanActivate {
 				console.log(error);
 				return of(null);
 			}),
-			map((session) => {
-				// The user isn't authenticated
+			switchMap((session) => {
+				// The user isn't authenticated, try reloading
 				if (session == null) {
-					// Send them to signin with a redirect to the previous URL
-					this.sessionService.setPreviousUrl(state.url);
-					this.router.navigate(['/signin']);
-					return false;
+					return this.sessionService.reloadSession();
 				}
-
-				return true;
+				return of(session);
+			}),
+			map((session) => {
+				return this.checkAccess(route, state, session);
 			})
 		);
+	}
+
+	checkAccess(route: ActivatedRouteSnapshot, state: RouterStateSnapshot, session: Session): boolean {
+		// The user still isn't authenticated
+		if (!this.authorizationService.isAuthenticated()) {
+			// Send them to signin with a redirect to the previous URL
+			this.sessionService.setPreviousUrl(state.url);
+			this.router.navigate(['/signin']);
+			return false;
+		}
+
+
+		if (!this.authorizationService.isAdmin()) {
+			// -----------------------------------------------------------
+			// Check the role requirements for the route
+			// -----------------------------------------------------------
+
+			// compile a list of roles that are missing
+			let requiredRoles = get(route, 'data.roles', ['user']);
+			let missingRoles: any[] = [];
+			requiredRoles.forEach( (role: any) => {
+				if (!this.authorizationService.hasRole(role)) {
+					missingRoles.push(role);
+				}
+			});
+
+			// If there are roles missing then we need to do something
+			if (missingRoles.length > 0) {
+
+				if (!this.authorizationService.hasRole(Role.USER)) {
+					// If the user is missing the user role, they are pending
+					if (state.url !== '/inactive-user') {
+						this.sessionService.setPreviousUrl(state.url);
+						this.router.navigate(['/inactive-user']);
```

Comment:
The desire with `access.component.ts` was to create a single component that would take route params indicating what the access problem was. So, we'd always route to the same single endpoint with different route params indicating if the issue was that the user was inactive, unauthorized, etc. You can look at `auth.interceptor.ts` to see an example of what I was thinking.

Reply:
This approach works great for the AuthInterceptor where the error details are dynamic and returned from the rest server, although I am not a fan of how all the text appears in the url.  What do you think would be the best way to approach handling more static errors like these?  I don't like the idea of the error text just being inline in the AuthGuard.  I was thinking about storing common static errors in an object in `access.component.ts` that can be used to look up the status/message based on the specified type.